[CI] CrossClustersQueryIT testProfile failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/zdb7vfq4wrmde/tests/:x-pack:plugin:esql:internalClusterTest/org.elasticsearch.xpack.esql.action.CrossClustersQueryIT/testProfile


**Reproduction line:**
```
./gradlew ':x-pack:plugin:esql:internalClusterTest' --tests "org.elasticsearch.xpack.esql.action.CrossClustersQueryIT.testProfile" -Dtests.seed=B25791012D271B46 -Dtests.locale=tr-TR -Dtests.timezone=Singapore -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
[Failure dashboard for `org.elasticsearch.xpack.esql.action.CrossClustersQueryIT#testProfile`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testProfile'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.action.CrossClustersQueryIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))


**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: <7>
     but: was <6>

  at __randomizedtesting.SeedInfo.seed([B25791012D271B46:E30D2484896AA36F]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.xpack.esql.action.CrossClustersQueryIT.testProfile(CrossClustersQueryIT.java:164)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

--- a/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeSourceHandler.java
+++ b/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeSourceHandler.java
@@ -12,7 +12,7 @@ import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.support.SubscribableListener;
 import org.elasticsearch.common.util.concurrent.AbstractRunnable;
 import org.elasticsearch.compute.data.Page;
-import org.elasticsearch.core.AbstractRefCounted;
+import org.elasticsearch.core.Releasable;
 import org.elasticsearch.tasks.TaskCancelledException;
 
 import java.util.concurrent.Executor;
@@ -27,18 +27,19 @@ import java.util.concurrent.atomic.AtomicReference;
  * @see #createExchangeSource()
  * @see #addRemoteSink(RemoteSink, int)
  */
-public final class ExchangeSourceHandler extends AbstractRefCounted {
+public final class ExchangeSourceHandler {
     private final ExchangeBuffer buffer;
     private final Executor fetchExecutor;
 
-    private final PendingInstances outstandingSinks = new PendingInstances();
-    private final PendingInstances outstandingSources = new PendingInstances();
+    private final PendingInstances outstandingSinks;
+    private final PendingInstances outstandingSources;
     private final AtomicReference<Exception> failure = new AtomicReference<>();
-    private final SubscribableListener<Void> completionFuture = new SubscribableListener<>();
 
     public ExchangeSourceHandler(int maxBufferSize, Executor fetchExecutor) {
         this.buffer = new ExchangeBuffer(maxBufferSize);
         this.fetchExecutor = fetchExecutor;
+        this.outstandingSinks = new PendingInstances(() -> buffer.finish(false));
+        this.outstandingSources = new PendingInstances(() -> buffer.finish(true));
     }
 
     private class LocalExchangeSource implements ExchangeSource {
@@ -76,9 +77,7 @@ public final class ExchangeSourceHandler extends AbstractRefCounted {
         public void finish() {
             if (finished == false) {
                 finished = true;
-                if (outstandingSources.finishInstance()) {
-                    buffer.finish(true);
-                }
+                outstandingSources.finishInstance();
             }
         }
 
@@ -205,9 +204,7 @@ public final class ExchangeSourceHandler extends AbstractRefCounted {
         void onSinkComplete() {
             if (finished == false) {
                 finished = true;
-                if (outstandingSinks.finishInstance()) {
-                    buffer.finish(false);
-                }
+                outstandingSinks.finishInstance();
             }
         }
     }
@@ -237,35 +234,36 @@ public final class ExchangeSourceHandler extends AbstractRefCounted {
         }
     }
 
-    @Override
-    protected void closeInternal() {
-        Exception error = failure.get();
-        if (error != null) {
-            completionFuture.onFailure(error);
-        } else {
-            completionFuture.onResponse(null);
-        }
-    }
-
     /**
-     * Add a listener, which will be notified when this exchange source handler is completed. An exchange source
-     * handler is consider completed when all exchange factories and sinks are completed and de-attached.
+     * Links this exchange source with an empty/dummy remote sink. The purpose of this is to prevent this exchange source from finishing
+     * until we have performed other async actions, such as linking actual remote sinks.
+     *
+     * @return a Releasable that should be called when the caller no longer needs to prevent the exchange source from completing.
      */
-    public void addCompletionListener(ActionListener<Void> listener) {
-        completionFuture.addListener(listener);
+    public Releasable addEmptySink() {
+        outstandingSinks.trackNewInstance();
+        return outstandingSinks::finishInstance;
     }
 
-    private final class PendingInstances {
+    private static class PendingInstances {
         private final AtomicInteger instances = new AtomicInteger();
+        private final Releasable onComplete;
+
+        PendingInstances(Releasable onComplete) {
+            this.onComplete = onComplete;
+        }
 
         void trackNewInstance() {
-            incRef();
-            instances.incrementAndGet();
+            int refs = instances.incrementAndGet();
+            assert refs > 0;
         }
 
-        boolean finishInstance() {
-            decRef();
-            return instances.decrementAndGet() == 0;
+        void finishInstance() {
+            int refs = instances.decrementAndGet();
+            assert refs >= 0;
+            if (refs == 0) {
+                onComplete.close();
+            }
         }
     }
 

--- a/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/exchange/ExchangeServiceTests.java
+++ b/x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/exchange/ExchangeServiceTests.java
@@ -54,8 +54,6 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Function;
 import java.util.function.Supplier;
@@ -95,11 +93,8 @@ public class ExchangeServiceTests extends ESTestCase {
         ExchangeSink sink1 = sinkExchanger.createExchangeSink();
         ExchangeSink sink2 = sinkExchanger.createExchangeSink();
         ExchangeSourceHandler sourceExchanger = new ExchangeSourceHandler(3, threadPool.executor(ESQL_TEST_EXECUTOR));
-        assertThat(sourceExchanger.refCount(), equalTo(1));
         ExchangeSource source = sourceExchanger.createExchangeSource();
-        assertThat(sourceExchanger.refCount(), equalTo(2));
         sourceExchanger.addRemoteSink(sinkExchanger::fetchPageAsync, 1);
-        assertThat(sourceExchanger.refCount(), equalTo(3));
         SubscribableListener<Void> waitForReading = source.waitForReading();
         assertFalse(waitForReading.isDone());
         assertNull(source.pollPage());
@@ -137,13 +132,7 @@ public class ExchangeServiceTests extends ESTestCase {
         sink2.finish();
         assertTrue(sink2.isFinished());
         assertTrue(source.isFinished());
-        assertBusy(() -> assertThat(sourceExchanger.refCount(), equalTo(2)));
         source.finish();
-        assertThat(sourceExchanger.refCount(), equalTo(1));
-        CountDownLatch latch = new CountDownLatch(1);
-        sourceExchanger.addCompletionListener(ActionListener.releasing(latch::countDown));
-        sourceExchanger.decRef();
-        assertTrue(latch.await(1, TimeUnit.SECONDS));
         ESTestCase.terminate(threadPool);
         for (Page page : pages) {
             page.releaseBlocks();

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java
@@ -19,7 +19,6 @@ import org.elasticsearch.action.support.ChannelActionListener;
 import org.elasticsearch.action.support.ContextPreservingActionListener;
 import org.elasticsearch.action.support.RefCountingListener;
 import org.elasticsearch.action.support.RefCountingRunnable;
-import org.elasticsearch.action.support.SubscribableListener;
 import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.util.BigArrays;
@@ -31,11 +30,9 @@ import org.elasticsearch.compute.operator.Driver;
 import org.elasticsearch.compute.operator.DriverProfile;
 import org.elasticsearch.compute.operator.DriverTaskRunner;
 import org.elasticsearch.compute.operator.ResponseHeadersCollector;
-import org.elasticsearch.compute.operator.exchange.ExchangeResponse;
 import org.elasticsearch.compute.operator.exchange.ExchangeService;
 import org.elasticsearch.compute.operator.exchange.ExchangeSinkHandler;
 import org.elasticsearch.compute.operator.exchange.ExchangeSourceHandler;
-import org.elasticsearch.compute.operator.exchange.RemoteSink;
 import org.elasticsearch.core.IOUtils;
 import org.elasticsearch.core.Releasable;
 import org.elasticsearch.core.Releasables;
@@ -184,11 +181,9 @@ public class ComputeService {
             transportService.getThreadPool().executor(ESQL_THREAD_POOL_NAME)
         );
         try (
-            Releasable ignored = exchangeSource::decRef;
+            Releasable ignored = exchangeSource.addEmptySink();
             RefCountingListener refs = new RefCountingListener(listener.map(unused -> new Result(collectedPages, collectedProfiles)))
         ) {
-            // wait until the source handler is completed
-            exchangeSource.addCompletionListener(refs.acquire());
             // run compute on the coordinator
             runCompute(
                 rootTask,
@@ -263,19 +258,6 @@ public class ComputeService {
         return remoteClusters;
     }
 
-    static final class EmptyRemoteSink implements RemoteSink {
-        final SubscribableListener<Void> future = new SubscribableListener<>();
-
-        @Override
-        public void fetchPageAsync(boolean allSourcesFinished, ActionListener<ExchangeResponse> listener) {
-            future.addListener(listener.map(ignored -> new ExchangeResponse(null, true)));
-        }
-
-        void finish() {
-            future.onResponse(null);
-        }
-    }
-
     private void startComputeOnDataNodes(
         String sessionId,
         String clusterAlias,
@@ -285,18 +267,17 @@ public class ComputeService {
         Set<String> concreteIndices,
         String[] originalIndices,
         ExchangeSourceHandler exchangeSource,
-        Supplier<ActionListener<ComputeResponse>> listener
+        Supplier<ActionListener<ComputeResponse>> listenerSupplier
     ) {
-        // Do not complete the exchange sources until we have linked all remote sinks
-        final EmptyRemoteSink emptyRemoteSink = new EmptyRemoteSink();
-        exchangeSource.addRemoteSink(emptyRemoteSink, 1);
+        // prevent the exchange source from completing until we have linked all remote sinks.
+        final Releasable emptySink = exchangeSource.addEmptySink();
         QueryBuilder requestFilter = PlannerUtils.requestFilter(dataNodePlan);
         lookupDataNodes(parentTask, clusterAlias, requestFilter, concreteIndices, originalIndices, ActionListener.wrap(dataNodes -> {
-            try (RefCountingRunnable refs = new RefCountingRunnable(emptyRemoteSink::finish)) {
+            try (RefCountingRunnable refs = new RefCountingRunnable(emptySink::close)) {
                 // For each target node, first open a remote exchange on the remote node, then link the exchange source to
                 // the new remote exchange sink, and initialize the computation on the target node via data-node-request.
                 for (DataNode node : dataNodes) {
-                    var dataNodeListener = ActionListener.releaseAfter(listener.get(), refs.acquire());
+                    var dataNodeListener = ActionListener.releaseAfter(listenerSupplier.get(), refs.acquire());
                     var queryPragmas = configuration.pragmas();
                     ExchangeService.openExchange(
                         transportService,
@@ -319,10 +300,7 @@ public class ComputeService {
                     );
                 }
             }
-        }, e -> {
-            emptyRemoteSink.finish();
-            listener.get().onFailure(e);
-        }));
+        }, e -> ActionListener.runAfter(listenerSupplier.get(), emptySink::close).onFailure(e)));
     }
 
     private void startComputeOnRemoteClusters(
@@ -334,10 +312,7 @@ public class ComputeService {
         List<RemoteCluster> clusters,
         Supplier<ActionListener<ComputeResponse>> listener
     ) {
-        // Do not complete the exchange sources until we have linked all remote sinks
-        final EmptyRemoteSink emptyRemoteSink = new EmptyRemoteSink();
-        exchangeSource.addRemoteSink(emptyRemoteSink, 1);
-        try (RefCountingRunnable refs = new RefCountingRunnable(emptyRemoteSink::finish)) {
+        try (RefCountingRunnable refs = new RefCountingRunnable(exchangeSource.addEmptySink()::close)) {
             for (RemoteCluster cluster : clusters) {
                 var targetNodeListener = ActionListener.releaseAfter(listener.get(), refs.acquire());
                 var queryPragmas = configuration.pragmas();
@@ -666,11 +641,7 @@ public class ComputeService {
             configuration.pragmas().exchangeBufferSize(),
             transportService.getThreadPool().executor(ESQL_THREAD_POOL_NAME)
         );
-        try (
-            Releasable ignored = exchangeSource::decRef;
-            RefCountingListener refs = new RefCountingListener(listener.map(unused -> new ComputeResponse(collectedProfiles)))
-        ) {
-            exchangeSource.addCompletionListener(refs.acquire());
+        try (RefCountingListener refs = new RefCountingListener(listener.map(unused -> new ComputeResponse(collectedProfiles)))) {
             exchangeSink.addCompletionListener(refs.acquire());
             PhysicalPlan coordinatorPlan = new ExchangeSinkExec(
                 plan.source(),

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java
@@ -208,6 +208,7 @@ public class ComputeService {
                     Set.of(localConcreteIndices.indices()),
                     localOriginalIndices.indices(),
                     exchangeSource,
+                    ActionListener.releaseAfter(refs.acquire(), exchangeSource.addEmptySink()),
                     () -> cancelOnFailure(rootTask, cancelled, refs.acquire()).map(response -> {
                         responseHeadersCollector.collect();
                         if (configuration.profile()) {
@@ -267,17 +268,16 @@ public class ComputeService {
         Set<String> concreteIndices,
         String[] originalIndices,
         ExchangeSourceHandler exchangeSource,
-        Supplier<ActionListener<ComputeResponse>> listenerSupplier
+        ActionListener<Void> parentListener,
+        Supplier<ActionListener<ComputeResponse>> dataNodeListenerSupplier
     ) {
-        // prevent the exchange source from completing until we have linked all remote sinks.
-        final Releasable emptySink = exchangeSource.addEmptySink();
         QueryBuilder requestFilter = PlannerUtils.requestFilter(dataNodePlan);
         lookupDataNodes(parentTask, clusterAlias, requestFilter, concreteIndices, originalIndices, ActionListener.wrap(dataNodes -> {
-            try (RefCountingRunnable refs = new RefCountingRunnable(emptySink::close)) {
+            try (RefCountingRunnable refs = new RefCountingRunnable(() -> parentListener.onResponse(null))) {
                 // For each target node, first open a remote exchange on the remote node, then link the exchange source to
                 // the new remote exchange sink, and initialize the computation on the target node via data-node-request.
                 for (DataNode node : dataNodes) {
-                    var dataNodeListener = ActionListener.releaseAfter(listenerSupplier.get(), refs.acquire());
+                    var dataNodeListener = ActionListener.releaseAfter(dataNodeListenerSupplier.get(), refs.acquire());
                     var queryPragmas = configuration.pragmas();
                     ExchangeService.openExchange(
                         transportService,
@@ -300,7 +300,7 @@ public class ComputeService {
                     );
                 }
             }
-        }, e -> ActionListener.runAfter(listenerSupplier.get(), emptySink::close).onFailure(e)));
+        }, parentListener::onFailure));
     }
 
     private void startComputeOnRemoteClusters(
@@ -673,6 +673,7 @@ public class ComputeService {
                 concreteIndices,
                 originalIndices,
                 exchangeSource,
+                ActionListener.releaseAfter(refs.acquire(), exchangeSource.addEmptySink()),
                 () -> cancelOnFailure(parentTask, cancelled, refs.acquire()).map(r -> {
                     responseHeadersCollector.collect();
                     if (configuration.profile()) {

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java
@@ -641,7 +641,10 @@ public class ComputeService {
             configuration.pragmas().exchangeBufferSize(),
             transportService.getThreadPool().executor(ESQL_THREAD_POOL_NAME)
         );
-        try (RefCountingListener refs = new RefCountingListener(listener.map(unused -> new ComputeResponse(collectedProfiles)))) {
+        try (
+            Releasable ignored = exchangeSource.addEmptySink();
+            RefCountingListener refs = new RefCountingListener(listener.map(unused -> new ComputeResponse(collectedProfiles)))
+        ) {
             exchangeSink.addCompletionListener(refs.acquire());
             PhysicalPlan coordinatorPlan = new ExchangeSinkExec(
                 plan.source(),

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java
@@ -412,7 +412,7 @@ public class CsvTests extends ESTestCase {
             }));
             return future.actionGet(TimeValue.timeValueSeconds(30));
         } finally {
-            Releasables.close(() -> Releasables.close(drivers), exchangeSource::decRef);
+            Releasables.close(() -> Releasables.close(drivers));
         }
     }
 

